symengine/sets.cpp

```diff
@@ -1283,6 +1292,109 @@ vec_basic Union::get_args() const
     return v;
 }
 
+Intersection::Intersection(const set_set &in)
+    : container_(in){SYMENGINE_ASSIGN_TYPEID()
+                         SYMENGINE_ASSERT(Intersection::is_canonical(in))}
+
+      hash_t Intersection::__hash__() const
+{
+    hash_t seed = SYMENGINE_INTERSECTION;
+    for (const auto &a : container_)
+        hash_combine<Basic>(seed, *a);
+    return seed;
+}
+
+bool Intersection::__eq__(const Basic &o) const
+{
+    if (is_a<Intersection>(o)) {
+        const Intersection &other = down_cast<const Intersection &>(o);
+        return unified_eq(container_, other.container_);
+    }
+    return false;
+}
+
+bool Intersection::is_canonical(const set_set &in)
+{
+    if (in.size() <= 1)
+        return false;
+    int count = 0;
+    for (const auto &s : in) {
+        if (is_a<FiniteSet>(*s)) {
+            count++;
+        }
+        if (count >= 2)
+            return false;
+    }
+    return true;
```

Comment:
Is this complete? For example if there are two `Interval`s they should get simplified automatically.

Reply:
It is even incorrect! We must allow two or more `FiniteSet`. For example:

```python
import sympy

x = sympy.Symbol("x")
y = sympy.Symbol("y")
xs = sympy.FiniteSet(x)
ys = sympy.FiniteSet(y)

xs.intersect(ys)
Out[1]: 
Intersection({x}, {y})
```

Reply:
Added a commit to fix this. This also resolved a previously unknown issue:

A unit test checked that {0, 1, x} ∩ {0, 1, 2} = {0, 1}. This is not true since we don't know if x is 2. Sympy keeps the original input here and symengine does that now after the fix.